tools/ob_admin/backup_tool/ob_admin_backup_validation_task.cpp

```diff
@@ -0,0 +1,4804 @@
+              STORAGE_LOG(WARN, "failed to get ls attr", K(ret), K(ls_id));
+            } else if (FALSE_IT(ls_attr->ls_type_
+                                = ObAdminBackupLSValidationAttr::ObAdminLSType::POST_CONSTRUCTED)) {
+            } else {
+              STORAGE_LOG(WARN, "post constructed logstream fixed", K(ret), K(ls_id));
+            }
+          } else {
+            STORAGE_LOG(WARN, "failed to get ls attr", K(ret), K(ls_id));
+          }
+        }
+        if (OB_FAIL(ret)) {
+          STORAGE_LOG(WARN, "failed to get ls attr", K(ret), K(ls_id));
+        } else if (OB_FAIL(inner_assign_meta_index_to_tablet_attr_(
+                       backup_set_id, ls_id, *sec_meta_index, backup_major_data_type_))) {
+          STORAGE_LOG(WARN, "failed to assign meta index to tablet attr", K(ret), K(ls_id),
+                      K(sec_meta_index->meta_key_));
+        }
+      }
+    }
+
+    FOREACH_X(meta_index, meta_index_list, OB_SUCC(ret))
+    {
+      if (OB_ISNULL(meta_index)) {
+        ret = OB_ERR_UNEXPECTED;
+        STORAGE_LOG(WARN, "meta index is null", K(ret));
+      } else if (!meta_index->meta_key_.is_valid()) {
+        ret = OB_ERR_UNEXPECTED;
+        STORAGE_LOG(WARN, "failed to check meta key", K(ret), K(meta_index->meta_key_));
+      } else if (backup::ObBackupMetaType::BACKUP_SSTABLE_META != meta_index->meta_key_.meta_type_
+                 && backup::ObBackupMetaType::BACKUP_TABLET_META
+                        != meta_index->meta_key_.meta_type_) {
+        ret = OB_ERR_UNEXPECTED;
+        STORAGE_LOG(WARN, "failed to check meta key", K(ret), K(meta_index->meta_key_));
+      } else if (meta_index->meta_key_.tablet_id_.is_ls_inner_tablet()) {
+        ret = OB_ERR_UNEXPECTED;
+        STORAGE_LOG(WARN, "unexpected ls inner tablet", K(ret), K(meta_index->meta_key_));
+      } else if (meta_index->backup_set_id_ != backup_set_id) {
+        ret = OB_ERR_UNEXPECTED;
+        STORAGE_LOG(WARN, "backup set id not match", K(ret), K(meta_index->backup_set_id_),
+                    K(backup_set_id));
+      } else {
+        ObLSID ls_id = meta_index->ls_id_;
+        ObAdminBackupLSValidationAttr *ls_attr = nullptr;
+        if (OB_FAIL(ctx_->get_ls_attr(backup_set_id, ls_id, ls_attr))) {
+          // should not have more ls here
+          STORAGE_LOG(WARN, "failed to get ls attr", K(ret), K(ls_id));
+        } else if (OB_FAIL(inner_assign_meta_index_to_tablet_attr_(
+                       backup_set_id, ls_id, *meta_index, backup_major_data_type_))) {
+          STORAGE_LOG(WARN, "failed to assign meta index to tablet attr", K(ret), K(ls_id),
+                      K(meta_index->meta_key_));
+        }
+      }
+    }
+  }
+  if (OB_NOT_NULL(ctx_) && OB_FAIL(ret)) {
+    ctx_->go_abort(full_path.get_ptr(), "major data info seems corrputed");
+  }
+  return ret;
+}
+int ObAdminBackupSetMetaValidationTask::inner_check_tablet_meta_index_(
```

Comment:
I cannot understand this function

Reply:
the func is to basically check if index meta cloud match with tablet record. use three offset to push forward, check if any more or less